### ISSUE TITLE
Configuring GMW_CONF_T_DATA for Chaos to reach all Flocklab nodes

### ISSUE THREAD

**Issue by swarth100** [ETHZ-TEC/Baloo#15]
Hi @romain-jacob, I have been playing extensively with Baloo+chaos over the past few days and I think I've encountered an issue which I managed to replicate with the following code (see below and on attached pastebins).

I think the currently maximum allowed values for `GMW_CONF_T_DATA` do not allow Chaos to correctly execute reaching and aggregating data from all nodes when run on Flocklab.

If my testing is correct currently a Baloo data slot has maximum size of ~128ms.

When interpreting the results from the A2:Synchotron paper (the flocklab tests), it really seems like an individual chaos round (full network flood + payload aggregation from each node) usually takes them between 200 and 300 ms to execute.

When testing on Flocklab, with the maximum ~128ms slot size I am never able to reliably see the payload from all nodes aggregated correctly into the final message which reaches the initiator.

I have been testing mechanisms to attempt to recover from this (attempt to execute multiple chaos round, keeping track locally of chaos flags and payload so they can be used again the round after), but overall this yields terrible latency (~2-5 times slower than A2:Synchotron to perform the same protocols).

Therefore I am wondering if:
1) There is a way to extend the maximum `GMW_CONF_T_DATA` size.
2) There is a way to have nodes "resume" an interrupted chaos flood. I can reassign flags and payloads, but my understanding of chaos is that it is always triggered by one node declared as `is_initiator`. When multiple nodes have the initiaton flag set, the floods seem to not occur correctly (am I testing badly)?
3) Does this data make sense? Is my testing completely wrong? If so would you just be able to give a small quick hint into what my biggest understanding flaw could be?

I have linked the code and results of running on Flocklab.

- - -
The code:

This is a minor modification of the `baloo-test-chaos` example, in rder to keep it as simple as possible.

Full [`baloo-test-chaos.c` file](https://pastebin.com/7uY2h7Zi).
Full [`project-conf.h` file](https://pastebin.com/tzu2jRwp).
Full [Flocklab results file](https://pastebin.com/Eseq9rcC) (initiator was node 10).

As I am testing on all 26 flocklab nodes, I modified the following:
```
#define NODE_LIST  \
 { 1,2,3,4,6,7,8,10,11,13,14,15,16,17,18,19,20,22,23,24,25,26,27,28,32,33 }
#define NUM_NODES                       26
#define CHAOS_CONF_PAYLOAD_LEN          4
```

I configured `GMW_CONF_T_DATA` rounded down to the maximum 10ms period:
```
#define GMW_CONF_T_DATA                 120000UL
```

Greater values (such as `130000UL`, `200000UL`) will not compile with error:
```
error: large integer implicitly truncated to unsigned type
```

When it comes to the code I am just looking at the payload. All nodes will wish to aggregate their payload every round:

```
static gmw_skip_event_t
src_on_slot_pre_callback(uint8_t slot_index,
                         uint16_t slot_assignee,
                         uint8_t* out_len,
                         uint8_t* out_payload,
                         uint8_t is_initiator,
                         uint8_t is_contention_slot)
{
  /* Always aggregate into the payload */
  *out_payload = 1;
  *out_len = 1;
  return GMW_EVT_SKIP_DEFAULT;
}
```

And aggregation just sets the relevant bit in the payload:
```
__attribute__((always_inline))
inline void chaos_set_payload_cb(uint8_t* chaos_payload_field,
                                 uint16_t node_index,
                                 uint8_t* payload)
{
  if(*payload) {
    chaos_payload_field[node_index / 8] |= (1 << (node_index % 8));
  }
}
```

Results printing is simply done by investigating the contents of the payload to see if all nodes have replied:
```
static gmw_repeat_event_t
src_on_slot_post_callback(uint8_t slot_index,
                          uint16_t slot_assignee,
                          uint8_t len,
                          uint8_t* payload,
                          uint8_t is_initiator,
                          uint8_t is_contention_slot,
                          gmw_pkt_event_t event)
{
  DEBUG_PRINT_INFO("payload: %u %u %u %u", payload[0], payload[1], payload[2], payload[3]);
  return GMW_EVT_REPEAT_DEFAULT;
}
```

A correct chaos round would have reached all 26 nodes, yielding an output (on initiator node with ID 10) of:
```
payload: 255 255 255 3
```

Whereas when testing on Flocklab I get results of the form:
```
1555065953.090825,10,10,r,[   0] payload: 128 128 129 3
[...]
1555065955.090949,10,10,r,[   0] payload: 128 1 48 0
[...]
1555065957.091816,10,10,r,[   0] payload: 132 8 0 1
[...]
1555065959.094163,10,10,r,[   0] payload: 128 3 16 0
[...]
1555065961.092170,10,10,r,[   0] payload: 128 2 49 0
[...]
```

Therefore never all results from all nodes are present

**Comment by romain-jacob** [comment]
Read [this page in the Wiki, about time management](https://github.com/ETHZ-TEC/Baloo/wiki/Baloo-Managing-time).

To address your problem, the "intended" way is to increase the time base for the slot time (from 500us to 1ms for example). However, this feature has not been extensively tested, so I cannot guarantee it will work smoothly.

If this crashes (please let me know...), a "simple" workaround would be to increase the slot time field in the control packet (from 8 to 16 bits), which will give you much larger slot time that what you will ever need... That will likely require some adaptation in the functions from `gmw-control.c` and it might not be so easy to hack.

Try increasing the time base first.

**Comment by swarth100** [comment]
Thank you so much for your prompt reply!

I've been trying various combinations of `GMW_CONF_SLOT_TIME_BASE` and `GMW_CONF_T_DATA`, but I still seem to have a hard time retrieving replies from all Flocklab nodes :/

```
#define GMW_CONF_SLOT_TIME_BASE           1000UL
#define GMW_CONF_T_DATA                   240000UL

---

#define GMW_CONF_SLOT_TIME_BASE           2000UL
#define GMW_CONF_T_DATA                   480000UL
```

I'm afraid I'm misconfiguring something completely.
Is there any other macro you would recommend adjusting?

**Comment by romain-jacob** [comment]
I think these settings should work. One thing I would recommend is to look with a logic analyzer (or on the Flocklab traces) if the slots indeed have the correct length (i.e.,  ~ 240 ms or 480 ms in your case). The `CHAOS_START_PIN` should give you that.

If slots have the expected length, then the problem likely lies within the Chaos implementation/configuration. If not, then that's a Baloo problem.

Let me know.

**Comment by swarth100** [comment]
So, this is slightly going into territory I've never really explored before, but simply from an overview of Flocklab,

Here is the trace of a variable [multi-round protocol](https://imgur.com/qBjbxzc) running on top of Baloo (still within the booting stage). It only uses the Glossy primitive. And as far as I understand all nodes are replying/receiving communications in their slots.

Here [is the trace](https://imgur.com/5tzed1t) when running a single 480ms Chaos round (with the above settings). From what I can see it seems like only the first flood is initiated.

Now I'm just gonna test the above code with 480ms slots on a brand new copy of Baloo. I am afraid I might have tampered with something that should not have been touched, so will just report the trace on a brand new copy.

---
EDIT:

Tested on a brand new clone of the latest Baloo repo.
I copied the `baloo-chaos-test.c` and `project-conf.h` above. So the aim is to simply reach out and aggregate a bit in the payload for every node in the network.

I configured a 480ms data slot length:
```
#define GMW_CONF_SLOT_TIME_BASE           2000UL
#define GMW_CONF_T_DATA                   480000UL
```

Had to comment out a line within `gmw-conf.h` to allow for compilation.
```
#ifndef GMW_CONF_SLOT_TIME_BASE
#define GMW_CONF_SLOT_TIME_BASE           500UL /* in ms */
#else
// #error "GMW_CONF_SLOT_TIME_BASE already defined!"
#endif
```

I do get serial results of the form:
```
1555249953.621443,10,10,r,[   0] payload: 132 0 0 3
1555249953.621443,10,10,r,[   0] time: 14, period: 2
1555249953.653604,10,10,r,[   0] round finished
1555249955.621448,10,10,r,[   0] payload: 128 2 48 0
1555249955.621448,10,10,r,[   0] time: 16, period: 2
1555249955.650302,10,10,r,[   0] round finished
1555249957.621522,10,10,r,[   0] payload: 132 0 48 0
1555249957.621522,10,10,r,[   0] time: 18, period: 2
1555249957.653696,10,10,r,[   0] round finished
1555249959.621540,10,10,r,[   0] payload: 128 8 0 1
1555249959.621540,10,10,r,[   0] time: 20, period: 2
1555249959.653695,10,10,r,[   0] round finished
1555249961.621495,10,10,r,[   0] payload: 128 2 48 0
1555249961.621495,10,10,r,[   0] time: 22, period: 2
1555249961.660095,10,10,r,[   0] round finished
1555249963.620729,10,10,r,[   0] payload: 128 2 48 0
1555249963.620729,10,10,r,[   0] time: 24, period: 2
1555249963.652699,10,10,r,[   0] round finished
1555249965.620807,10,10,r,[   0] payload: 132 0 36 1
```

And a[ Pin output](https://imgur.com/GDFzrj9) very similar to the one above.

---
EDIT 2:

So by looking at which nodes are replying within the payload. I've mapped payload indexes (actually 2^index) to node_ids:
```
# BYTE 0        index to node_id
1        1
2        2
4        3
8        4
16       6
32       7
64       8
128      10

# BYTE 1        index to node_id
1        11
2        13
4        14
8        15
16       16
32       17
64       18
128      19

# BYTE 2        index to node_id
1        20
2        22
4        23
8        24
16       25
32       26
64       27
128      28

# BYTE 3        index to node_id
1        32
2        33
 { 1,2,3,4,6,7,8,10,11,13,14,15,16,17,18,19,20,22,23,24,25,26,27,28,32,33 }
```

By looking at the [Flocklab Connectivity map](https://user.flocklab.ethz.ch/user/topology.php), Node 10 (the initiator) should be connected to:
```
3, 11, 15, 25, 26, 31, 32
```

We are not testing on node `31`, but this means that just by looking at the one-hop neighbours the maximal bitmap we could achieve is:

```
1000 0100     0000 1001     0001 0000     0000 0001
```

And if we look at the payloads received above:
```
1555249953.621443,10,10,r,[   0] payload: 132 0 0 3
1555249955.621448,10,10,r,[   0] payload: 128 2 48 0
1555249957.621522,10,10,r,[   0] payload: 132 0 48 0
1555249959.621540,10,10,r,[   0] payload: 128 8 0 1
1555249961.621495,10,10,r,[   0] payload: 128 2 48 0
1555249963.620729,10,10,r,[   0] payload: 128 2 48 0
1555249965.620807,10,10,r,[   0] payload: 132 0 36 1
```

They can be rewritten as:
```
REFERENCE:    1000 0100     0000 1001     0001 0000     0000 0001

TIME 14       1000 0100     0000 0000     0000 0000     0000 0011
TIME 16       1000 0000     0000 0010     0011 0000     0000 0000
TIME 18       1000 0100     0000 0000     0011 0000     0000 0000
TIME 20       1000 0000     0000 1000     0000 0000     0000 0001
TIME 22       1000 0000     0000 0010     0011 0000     0000 0000
TIME 24       1000 0000     0000 0010     0011 0000     0000 0000
TIME 26       1000 0100     0000 0000     0010 0100     0000 0001
```

Which looks very similar to the payload we would have when communicating to 10's one-hop neighbours.

---
So, I just don't want to waste your time, I was just trying to build protocols on top of Chaos and I feel like there is something I'm missing to get it to work completely right :(

**Comment by swarth100** [comment]
I was just wondering if you had a second over the past few days to look into this.

I am currently tring to forcefully aggregate the contents of `chaos_payload` and `&CHAOS_PAYLOAD_FIELD` at every iteration as I feel like they do not always contain the same data.

Though I investigate most of this via print statements which might simply slow down the protocol making it yield incorrect results :/

**Comment by romain-jacob** [comment]
No, I did not. I will try something out today on Flocklab (e.g., each node putting its node ID in the packet). I let you know how it works for me.

**Comment by swarth100** [comment]
Thank you so much!

So due to scalability problems of the packet size on large networks, I am actually trying to aggregate a single bit (instead of a full 1 or 2 bytes) from each node into a shared Chaos packet

Nodes select the bit corresponding to their node_index

---
To implement one-bit payloads for each node,
I have tried to change the default `baloo-test-chaos.c` implementation as little as possible:

**`project-conf.h`**:
```
#define NODE_LIST  \
 { 1,2,3,4,6,7,8,10,11,13,14,15,16,17,18,19,20,22,23,24,25,26,27,28,32,33 }
#define NUM_NODES                       26

#define GMW_CONF_SLOT_TIME_BASE           2000UL
#define GMW_CONF_T_DATA                   480000UL

#define CHAOS_CONF_NUM_NODES            NUM_NODES
#define CHAOS_CONF_NODE_ID_MAPPING      NODE_LIST
#define CHAOS_CONF_PAYLOAD_LEN          4
#define CHAOS_CONF_SET_CUSTOM_PAYLOAD   1
#define CHAOS_CONF_SHARED_PAYLOAD       1

[...]
```

**`baloo-test-chaos.c`**:
```
__attribute__((always_inline))
inline void chaos_set_payload_cb(uint8_t* chaos_payload_field,
                                 uint16_t node_index,
                                 uint8_t* payload)
{
  if(*payload) {
    chaos_payload_field[node_index / 8] |= (1 << (node_index % 8));
  }
}

/***********/

static gmw_skip_event_t
src_on_slot_pre_callback(uint8_t slot_index,
                         uint16_t slot_assignee,
                         uint8_t* out_len,
                         uint8_t* out_payload,
                         uint8_t is_initiator,
                         uint8_t is_contention_slot)
{
  /* Always aggregate into the payload */
  *out_payload = 1;
  *out_len = CHAOS_CONF_PAYLOAD_LEN;
  return GMW_EVT_SKIP_DEFAULT;
}

/***********/

static gmw_repeat_event_t
src_on_slot_post_callback(uint8_t slot_index,
                          uint16_t slot_assignee,
                          uint8_t len,
                          uint8_t* payload,
                          uint8_t is_initiator,
                          uint8_t is_contention_slot,
                          gmw_pkt_event_t event)
{
  DEBUG_PRINT_INFO("payload: %u %u %u %u", payload[0], payload[1], payload[2], payload[3]);
  return GMW_EVT_REPEAT_DEFAULT;
}
```

Though I always seem to get the incorrect results outlined in the earlier remarks to this issue.

I have seen you guys have worked a lot with Chaos+Baloo so I really was just wondering what configuration macro I'm missing, or what errors/incorrect assumptions I'm making with Chaos or Shared Chaos payloads.

**Comment by swarth100** [comment]
I have also tried aggregating one byte of information per node without using a shared payload and I get a very similar problem:

**`project-conf.h`**:
```
#define NODE_LIST  \
 { 1,2,3,4,6,7,8,10,11,13,14,15,16,17,18,19,20,22,23,24,25,26,27,28,32,33 }
#define NUM_NODES                       26

#define GMW_CONF_SLOT_TIME_BASE           2000UL
#define GMW_CONF_T_DATA                   480000UL

#define CHAOS_CONF_NUM_NODES            NUM_NODES
#define CHAOS_CONF_NODE_ID_MAPPING      NODE_LIST
#define CHAOS_CONF_PAYLOAD_LEN          NUM_NODES

/* CHAOS_CONF_SET_CUSTOM_PAYLOAD and CHAOS_CONF_SHARED_PAYLOAD are left undefined*/
```

**`baloo-test-chaos.c`**:
```
static gmw_skip_event_t
src_on_slot_pre_callback(uint8_t slot_index,
                         uint16_t slot_assignee,
                         uint8_t* out_len,
                         uint8_t* out_payload,
                         uint8_t is_initiator,
                         uint8_t is_contention_slot)
{
  /* Always aggregate into the payload */
  *out_payload = 1;
  *out_len = 1;
  return GMW_EVT_SKIP_DEFAULT;
}

/************/
static gmw_repeat_event_t
src_on_slot_post_callback(uint8_t slot_index,
                          uint16_t slot_assignee,
                          uint8_t len,
                          uint8_t* payload,
                          uint8_t is_initiator,
                          uint8_t is_contention_slot,
                          gmw_pkt_event_t event)
{
  DEBUG_PRINT_INFO("payload: %u %u %u %u %u %u %u %u %u %u %u %u %u %u %u %u",
    payload[0], payload[1], payload[2], payload[3], 
    payload[4], payload[5], payload[6], payload[7], 
    payload[8], payload[9], payload[10], payload[11], 
    payload[12], payload[13], payload[14], payload[15]
  );
  return GMW_EVT_REPEAT_DEFAULT;
}
```

And I obtain the following results on the initiator:
```
1556188190.281637,10,10,r,[   0] payload: 0 0 0 0 0 0 0 1 1 0 0 0 0 0 0 0
1556188190.281637,10,10,r,[   0] time: 166, period: 2
1556188190.314253,10,10,r,[   0] round finished
1556188192.281656,10,10,r,[   0] payload: 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 0
1556188192.281656,10,10,r,[   0] time: 168, period: 2
1556188192.320111,10,10,r,[   0] round finished
1556188194.281657,10,10,r,[   0] payload: 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 0
1556188194.281657,10,10,r,[   0] time: 170, period: 2
1556188194.320159,10,10,r,[   0] round finished
1556188196.281685,10,10,r,[   0] payload: 0 0 0 0 0 0 0 1 1 0 0 0 0 0 0 0
1556188196.281685,10,10,r,[   0] time: 172, period: 2
1556188196.320242,10,10,r,[   0] round finished
1556188198.281656,10,10,r,[   0] payload: 0 0 0 0 0 0 0 1 1 0 0 0 0 0 0 0
1556188198.281656,10,10,r,[   0] time: 174, period: 2
1556188198.320256,10,10,r,[   0] round finished
1556188200.281711,10,10,r,[   0] payload: 0 0 1 0 0 0 0 1 1 0 0 0 0 0 0 0
1556188200.281711,10,10,r,[   0] time: 176, period: 2
1556188200.314291,10,10,r,[   0] round finished
1556188202.282572,10,10,r,[   0] payload: 0 0 0 0 0 0 0 1 1 0 0 0 0 0 0 0
```

When I would be expecting a full payload of ones

**Comment by romain-jacob** [comment]
Okay, I think I found the problem. Chaos stops because the maximal number of retransmissions is reached. By default, `GMW_CONF_TX_CNT_DATA` is used, and its value is only 3. So with only 3 retransmissions per node, it's no surprise that the Chaos flood does not complete.
That's an important point I missed in the description of "how-to-use-Chaos", since I did not use it too much myself... I will correct that asap.

However, even with setting `GMW_CONF_TX_CNT_DATA` to the maximum of 255, the Chaos flood still does not complete on Flocklab. The problem looks like the flood stalls (i.e., nodes don't learn anything new, so they stop). I am not too sure anymore if this is the intended behavior of the primitive... Here is a typical Chaos flood I get:

![image](https://user-images.githubusercontent.com/32954999/56735757-a8555300-6766-11e9-9192-37c452b8f505.png)

Solving that issue would require to have a closer look into the Chaos implementation, which I don't have time to do now. But I feel that's where you want to look.

On the bright side, I tested the extension of the time base, which seems to work fine...
